{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "88eb9d5f",
   "metadata": {},
   "source": [
    "## Check Coverage of Schema properties by repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c0e9b224",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import requests\n",
    "from datetime import datetime\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b3571f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_coverage_table(sourcelist,propertylist,api):\n",
    "    ## classify the propertylist:\n",
    "    if 'name' in propertylist:\n",
    "        listtype = \"MR\"\n",
    "        if 'version' in propertylist:\n",
    "            listtype = 'MRO'\n",
    "    else:\n",
    "        listtype = \"O\"\n",
    "    ## Generate base df\n",
    "    print('Generating base df')\n",
    "    allist = []\n",
    "    if api == 'prod':\n",
    "        baseurl = 'api.data.niaid.nih.gov/v1/'\n",
    "    elif api == 'staging':\n",
    "        baseurl = 'api-staging.data.niaid.nih.gov/v1/'\n",
    "    for eachproperty in propertylist:\n",
    "        api_base_call = f'https://{baseurl}query?q=_exists_:{eachproperty}'\n",
    "        r = requests.get(api_base_call)\n",
    "        tmpdict = json.loads(r.text)\n",
    "        totals = tmpdict['total']\n",
    "        allist.append({'property':eachproperty,'All':totals})\n",
    "    alldf = pd.DataFrame(allist)\n",
    "    ## Append all other coverage data\n",
    "    totaldf = alldf.copy()\n",
    "    print('now fetching source coverage')\n",
    "    for eachsource in sourcelist:\n",
    "        sourcetotallist = []\n",
    "        print(\"now fetching for: \",eachsource)\n",
    "        for eachproperty in propertylist:\n",
    "            print(\"checking: \",eachproperty)\n",
    "            api_base_call = f'https://{baseurl}query?q=_exists_:{eachproperty} AND includedInDataCatalog.name:\"{eachsource}\"'\n",
    "            r = requests.get(api_base_call)\n",
    "            tmpdict = json.loads(r.text)\n",
    "            totals = tmpdict['total']\n",
    "            sourcetotallist.append({'property':eachproperty,f'{eachsource}':totals})\n",
    "        sourcedf = pd.DataFrame(sourcetotallist)\n",
    "        #print(len(sourcedf),' properties obtained for ',eachsource)\n",
    "        totaldf = totaldf.merge(sourcedf,on='property',how='left')\n",
    "        #print(totaldf.head(n=1))\n",
    "    ## Export the dataframe\n",
    "    now = datetime.now() # current date and time\n",
    "    date_info = now.strftime(\"%Y-%m-%d\")\n",
    "    totaldf.to_csv(os.path.join('coverage',f\"{date_info}_{api}_schema_coverage({listtype}).tsv\"),sep='\\t',header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "29e33a03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['name', 'description', 'author', 'author.name', 'author.givenName', 'author.familyName', 'url', 'measurementTechnique', 'measurementTechnique.name', 'includedInDataCatalog', 'includedInDataCatalog.name', 'includedIndataCatalog.archivedAt', 'distribution', 'distribution.contentUrl', 'distribution.dateModified', 'funding', 'funding.funder.name', 'funding.identifier', 'date', 'dateCreated', 'dateModified', 'datePublished', 'citedBy', 'doi', 'infectiousAgent', 'healthCondition', 'species', 'variableMeasured', 'citation', 'citation.pmid', 'citation.doi', 'conditionsOfAccess', 'isBasedOn', 'keywords', 'license', 'sdPublisher', 'spatialCoverage', 'temporalCoverage', 'topicCategory', 'identifier', 'usageInfo', 'interactionStatistic']\n"
     ]
    }
   ],
   "source": [
    "propertylistfile = os.path.join('data','propertylist.txt')\n",
    "propertylist = []\n",
    "with open(propertylistfile,'r') as inputfile:\n",
    "    for line in inputfile:\n",
    "        propertylist.append(line.strip())\n",
    "print(propertylist)\n",
    "\n",
    "sourcelistfile = os.path.join('data','sourcelist.txt')\n",
    "sourcelist = []\n",
    "with open(sourcelistfile,'r') as srcfile:\n",
    "    for line in srcfile:\n",
    "        sourcelist.append(line.strip())\n",
    "\n",
    "optionalpropsfile = os.path.join('data','optionallist.txt')\n",
    "optionallist = []\n",
    "with open(optionalpropsfile,'r') as optfile:\n",
    "    for line in optfile:\n",
    "        optionallist.append(line.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "822475ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['isRelatedTo', 'isSimilarTo', 'isBasisFor', 'nctid', 'abstract', 'aggregateRating', 'creator', 'hasPart', 'inLanguage', 'isAccessibleForFree', 'isPartOf', 'version', 'alternateName', 'mainEntityOfPage', 'sameAs', 'relationship', 'sourceOrganization.name']\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "print(optionallist)\n",
    "print(len(optionallist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "48aa89c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating base df\n",
      "now fetching source coverage\n",
      "now fetching for:  AccessClinicalData@NIAID\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  biotools\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  ClinEpiDB\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Data+Discovery+Engine\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Dryad+Digital+Repository\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Figshare\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Flow+Repository\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Harvard+Dataverse\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  HuBMAP\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Human+Cell+Atlas\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  ImmPort\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  ImmuneSpace\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  LINCS\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  MalariaGEN\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  MassIVE\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Mendeley\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  MicrobiomeDB\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  NCBI+BioProject\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  NCBI+GEO\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  NCBI+SRA\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  NDEx\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  NICHD+DASH\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Omics+Discovery+Index+(OmicsDI)\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Qiita\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  RADx+Data+Hub\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  ReframeDB\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  The+Database+of+Genotypes+and+Phenotypes\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  VDJServer\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  VEuPathDB\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Vivli\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "now fetching for:  Zenodo\n",
      "checking:  isRelatedTo\n",
      "checking:  isSimilarTo\n",
      "checking:  isBasisFor\n",
      "checking:  nctid\n",
      "checking:  abstract\n",
      "checking:  aggregateRating\n",
      "checking:  creator\n",
      "checking:  hasPart\n",
      "checking:  inLanguage\n",
      "checking:  isAccessibleForFree\n",
      "checking:  isPartOf\n",
      "checking:  version\n",
      "checking:  alternateName\n",
      "checking:  mainEntityOfPage\n",
      "checking:  sameAs\n",
      "checking:  relationship\n",
      "checking:  sourceOrganization.name\n",
      "CPU times: total: 37.4 s\n",
      "Wall time: 4min 16s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#api = 'staging'\n",
    "api = 'prod'\n",
    "#sourcelist = [\"MalariaGEN\",\"Omics+Discovery+Index+(OmicsDI)\"]\n",
    "generate_coverage_table(sourcelist,optionallist,api)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69e13bd5",
   "metadata": {},
   "source": [
    "## Investigate missing metadata\n",
    "\n",
    "1. Check for patterns of missing metadata \n",
    "    * for example, if there is an sdPublisher in OMICS-DI in particular for which the metadata went missing\n",
    "    * \"https://api.data.niaid.nih.gov/v1/query?q=-_exists_:description%20AND%20includedInDataCatalog.name:%22Omics+Discovery+Index+(OmicsDI)%22&fields=sdPublisher.name&fetch_all=true\"\n",
    "\n",
    "2. If necessary, check if the records missing metadata in Staging have values in Production"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17b38817",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36136e54",
   "metadata": {},
   "outputs": [],
   "source": [
    "#propertylist = ['conditionsOfAccess','description','dateCreated','datePublished']\n",
    "propertylist = ['conditionsOfAccess']\n",
    "#api = 'staging'\n",
    "api = 'prod'\n",
    "source = 'Omics+Discovery+Index+(OmicsDI)'\n",
    "field = 'sdPublisher.name'\n",
    "\n",
    "allresults = pd.DataFrame(columns=['propname','_id','_score','sdPublisher'])\n",
    "\n",
    "if api == 'prod':\n",
    "    baseurl = 'https://api.data.niaid.nih.gov/v1/query?q='\n",
    "elif api == 'staging':\n",
    "    baseurl = 'https://api-staging.data.niaid.nih.gov/v1/query?q='\n",
    "for eachproperty in propertylist:\n",
    "    api_base_call = f'{baseurl}includedInDataCatalog.name:\"{source}\"+AND+-_exists_:{eachproperty}&fields={field}&fetch_all=true'\n",
    "    r = requests.get(api_base_call)\n",
    "    tmpdict = json.loads(r.text)\n",
    "    totals = tmpdict['total']\n",
    "    tmpdf = pd.DataFrame(tmpdict['hits'])\n",
    "    tmpdf['propname'] = [eachproperty for x in tmpdf['_score']]\n",
    "    allresults = pd.concat((allresults,tmpdf),ignore_index=True)\n",
    "    if totals>=500:\n",
    "        i=0\n",
    "        maxscrolls = math.ceil(tmpdict['total']/500)\n",
    "        scroll_id = tmpdict['_scroll_id']\n",
    "        while i < maxscrolls:\n",
    "            try:\n",
    "                r2 = requests.get(f'https://{baseurl}query?scroll_id={scroll_id}')\n",
    "                tmp = json.loads(r2.text)\n",
    "                scroll_id = tmp['_scroll_id']\n",
    "                tmpdf = pd.DataFrame(tmp['hits'])\n",
    "                tmpdf['propname'] = [eachproperty for x in tmpdf['_score']]\n",
    "                allresults = pd.concat((allresults,tmpdf),ignore_index=True)\n",
    "                i=i+1\n",
    "            except:\n",
    "                break\n",
    "                \n",
    "print(allresults.head(n=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d326879f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(allresults))\n",
    "allresults['sdPubName'] = [x['name'] for x in allresults['sdPublisher'].tolist()]\n",
    "print(len(allresults.loc[allresults['_ignored'].isna()]))\n",
    "print(allresults.loc[~allresults['_ignored'].isna()].head(n=2))\n",
    "#cleanresults = allresults[['propname','_id','sdPubName']].copy()\n",
    "#cleanresults.drop_duplicates(keep='first',inplace=True)\n",
    "#print(cleanresults.head(n=2))\n",
    "#print(len(cleanresults))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4bce612",
   "metadata": {},
   "outputs": [],
   "source": [
    "#staging_histogram = cleanresults.groupby(['propname','sdPubName']).size().reset_index(name='staging_counts')\n",
    "print(staging_histogram.head(n=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59896e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(cleanresults))\n",
    "prod_histogram = cleanresults.groupby(['propname','sdPubName']).size().reset_index(name='prod_counts')\n",
    "print(prod_histogram)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d643eb91",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(cleanresults.loc[cleanresults['sdPubName']=='bioimages'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87386c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison = staging_histogram.merge(prod_histogram,on=['propname','sdPubName'],how='outer')\n",
    "print(comparison.head(n=2))\n",
    "comparison.to_csv(os.path.join('coverage','2025-07-10-stage_prod_missing_comparison.tsv'),sep='\\t',header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9142b3b",
   "metadata": {},
   "source": [
    "## Test function parts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf9f7a58",
   "metadata": {},
   "outputs": [],
   "source": [
    "example_call = 'https://api.data.niaid.nih.gov/v1/query?q=_exists_:isRelatedTo'\n",
    "#example_call = 'https://api.data.niaid.nih.gov/v1/query?q=_exists_:isRelatedTo AND includedInDataCatalog.name:\"Zenodo\"'\n",
    "\n",
    "r = requests.get(example_call)\n",
    "tmpdict = json.loads(r.text)\n",
    "print(tmpdict['total'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ec951fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "allist = []\n",
    "for eachproperty in optionallist:\n",
    "    api_base_call = f'https://api.data.niaid.nih.gov/v1/query?q=_exists_:{eachproperty}'\n",
    "    r = requests.get(api_base_call)\n",
    "    tmpdict = json.loads(r.text)\n",
    "    totals = tmpdict['total']\n",
    "    allist.append({'property':eachproperty,'All':totals})\n",
    "print(allist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6ba9311",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Generate the zenodo table as default\n",
    "\n",
    "zenodolist = []\n",
    "for eachproperty in propertylist:\n",
    "    api_base_call = f'https://api.data.niaid.nih.gov/v1/query?q=_exists_:{eachproperty} AND includedInDataCatalog.name:\"Zenodo\"'\n",
    "    r = requests.get(api_base_call)\n",
    "    tmpdict = json.loads(r.text)\n",
    "    totals = tmpdict['total']\n",
    "    zenodolist.append({'property':eachproperty,'Zenodo':totals})\n",
    "\n",
    "zenododf = pd.DataFrame(zenodolist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "411836a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(zenododf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb9d31a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "totaldf = zenododf.copy()\n",
    "for eachsource in sourcelist:\n",
    "    sourcetotallist = []\n",
    "    print(eachsource)\n",
    "    for eachproperty in propertylist:\n",
    "        api_base_call = f'https://api.data.niaid.nih.gov/v1/query?q=_exists_:{eachproperty} AND includedInDataCatalog.name:\"{eachsource}\"'\n",
    "        r = requests.get(api_base_call)\n",
    "        tmpdict = json.loads(r.text)\n",
    "        totals = tmpdict['total']\n",
    "        sourcetotallist.append({'property':eachproperty,f'{eachsource}':totals})\n",
    "    sourcedf = pd.DataFrame(sourcetotallist)\n",
    "    totaldf = totaldf.merge(sourcedf,on='property',how='left')\n",
    "    print(totaldf.head(n=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "373f6ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(totaldf.head(n=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c3829b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "totaldf.to_csv('schema_coverage_2023.10.11.txt',sep='\\t',header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91106f37",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "mendeleylist = []\n",
    "for eachproperty in propertylist:\n",
    "    api_base_call = f'https://api.data.niaid.nih.gov/v1/query?q=_exists_:{eachproperty} AND includedInDataCatalog.name:\"Mendeley\"'\n",
    "    r = requests.get(api_base_call)\n",
    "    tmpdict = json.loads(r.text)\n",
    "    totals = tmpdict['total']\n",
    "    mendeleylist.append({'property':eachproperty,'Zenodo':totals})\n",
    "\n",
    "mendeleydf = pd.DataFrame(mendeleylist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a2b7475",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(mendeleydf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c77f77b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc3c083e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfb36cb2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
